src/interp/interp-wasi.cc

```diff
@@ -0,0 +1,250 @@
+  }
+
+  uvwasi_s* uvwasi;
+  // The memory accociated with the instance.  Looked up once on startup
+  // and cached here.
+  Memory* memory;
+};
+
+std::unordered_map<Instance*, WasiInstance*> wasiInstances;
+
+// TODO(sbc): Auto-generate this.
+
+#define WASI_CALLBACK(NAME)                                          \
+  Result NAME(Thread& thread, const Values& params, Values& results, \
+              Trap::Ptr* trap) {                                     \
+    Instance* instance = thread.GetCallerInstance();                 \
+    assert(instance);                                                \
+    WasiInstance* wasi_instance = wasiInstances[instance];           \
+    if (wasi_instance->trace_stream) {                               \
+      wasi_instance->trace_stream->Writef(                           \
+          ">>> running wasi function \"%s\":\n", #NAME);             \
+    }                                                                \
+    return wasi_instance->NAME(params, results, trap);               \
+  }
+
+WASI_CALLBACK(proc_exit);
+WASI_CALLBACK(fd_write);
+
+}  // namespace
+
+namespace wabt {
+namespace interp {
+
+Result WasiBindImports(const Module::Ptr& module,
+                       RefVec& imports,
+                       Stream* stream,
+                       Stream* trace_stream) {
+  Store* store = module.store();
+  for (auto&& import : module->desc().imports) {
+    if (import.type.type->kind != ExternKind::Func) {
+      stream->Writef("wasi error: invalid import type: %s\n",
+                     import.type.name.c_str());
+      return Result::Error;
+    }
+
+    if (import.type.module != "wasi_snapshot_preview1") {
+      stream->Writef("wasi error: unknown module import: %s\n",
+                     import.type.module.c_str());
+      return Result::Error;
+    }
+
+    auto func_type = *cast<FuncType>(import.type.type.get());
+    auto import_name = StringPrintf("%s.%s", import.type.module.c_str(),
+                                    import.type.name.c_str());
+    HostFunc::Ptr host_func;
+
+    // TODO(sbc): Auto-generate this.
+    if (import.type.name == "proc_exit") {
+      host_func = HostFunc::New(*store, func_type, proc_exit);
+    } else if (import.type.name == "fd_write") {
+      host_func = HostFunc::New(*store, func_type, fd_write);
+    } else {
+      stream->Writef("unknown wasi_snapshot_preview1 import: %s\n",
+                     import.type.name.c_str());
+      return Result::Error;
+    }
+    imports.push_back(host_func.ref());
+  }
+
+  return Result::Ok;
+}
+
+Result WasiRunStart(const Instance::Ptr& instance,
+                    uvwasi_s* uvwasi,
+                    Stream* stream,
+                    Stream* trace_stream) {
+  Store* store = instance.store();
+  auto module = store->UnsafeGet<Module>(instance->module());
+  auto&& module_desc = module->desc();
+
+  Func::Ptr start;
+  Memory::Ptr memory;
+  for (auto&& export_ : module_desc.exports) {
```

Comment:
Seems like we may want a convenient API for this in interp too.

Reply:
You mean something like a `getExportByName` function?

Reply:
Yeah. It'll be slightly less efficient (since it has to look through twice), but probably not a big deal for most users.